### PR TITLE
[RFC] TOOLS/lua/autoload: add ignore_patterns option

### PR DESCRIPTION
**Pull request by oficsu**
Why? Sometimes you want to ignore thumbnails, auto-generated backups or just unwanted files. And here is a non-exhaustive list of real-world examples:
- user backup files: `%.bak%.mp4$` or `^bak-`;
- telegram-desktop chat export thumbnails: `_thumb%.jpg$`;
- a krita graphics editor backup suffix: `^~`

But it's an RFC because... there is little point in a single pattern. It would be better if we could configure a list of patterns. But... Should we just split by commas or write a custom parser for `ignore_pattern` that allows to escape commas? Maybe something else?

### CODE REVIEW

**Comment by oficsu** [comment]
Now the option is named `ignore_patterns` and accepts a comma-separated list. The function `Split` has been replaced with a more advanced version that allows not just split by commas, but also escapes commas

A little about my design choice. Since `%,` is a never valid lua pattern and user will never try to create such patterns, I decided not to require escaping of lua patterns twice. For example, option `%,` is a pattern `,` that matches just a single literal comma. And option `%%%,` is a pattern `%%,` that matches single literal `%` then single literal comma. Personally, I think it's much better than double escaping

I didn't see any unit test for lua scripts, so I tested several cases out of mpv source with [luassert](https://github.com/lunarmodules/luassert):
```lua
local assert_equal = require 'luassert'.are.same

-- Empty lists
assert_equal(Split(""), Set{})
assert_equal(Split(","), Set{})
assert_equal(Split(",,"), Set{})

-- Normal Lists
assert_equal(Split("42"), Set{"42"})
assert_equal(Split("1,2"), Set{"1", "2"})
assert_equal(Split("Hello, World!"), Set{"Hello", " World!"})

-- Escaped commas
assert_equal(Split("%,"), Set{","})
assert_equal(Split(",%,"), Set{","})
assert_equal(Split("%,,"), Set{","})
assert_equal(Split("%,%,"), Set{",,"})
assert_equal(Split("-%,-"), Set{"-,-"})
assert_equal(Split("-%%%,-"), Set{"-%%,-"})

-- Comma is not escaped, forward escape sequences as is
assert_equal(Split("-%%,-"), Set{"-%%", "-"})
assert_equal(Split("-%%%%,-"), Set{"-%%%%", "-"})

-- A real-world example
assert_equal(Split("^~,^bak-,%.bak%.mp4$"), Set{"^~", "^bak-", "%.bak%.mp4$"})
```

**Comment by oficsu** [comment]
There was a little bunch of style-related fixes (indentation, extra `;`) and removing of `NEVERMATCHPATTERN` (can be replaced with just an empty list now). Since the code is essentially unchanged, I've just force-pushed the last commit

**Comment by guidocella** [comment]
Is there a reason to this in autoload.lua instead of ignoring patterns in any playlist mpv loads? This can already be done globally with a script like https://github.com/mpv-player/mpv/issues/2595#issuecomment-1454746798

**Comment by oficsu** [comment]
@guidocella,

> Is there a reason to this in autoload.lua instead of ignoring patterns in any playlist mpv loads?
1) I'm not sure about performance aspects, but I don't think it's a good idea to load (especially big) playlists with autoload.lua _only to remove_ something from it later with another script. Even if there are no performance issues, there can be some glitches and unpredictable bugs caused by the asynchronous nature of `observe_property`
2) It doesn't work. Autoload.lua script already cares about the explicitly opened file and doesn't remove it from playlist even if this file is excluded by script options. Another script will have to reimplement these checks
3) UX. A configurable comma-separated list of patterns is more convenient for the end users than manual scripting every time you want to exclude something. It's always better to have a domain-specific language for simple tasks, and it's always worse to force using of general-purpose language
4) Yes, eventually, instead of improving autoload.lua, you can write a new sufficiently flexible script that does not have some of these issues, but then it will... reimplement much of existing code
5) In general, the only one autoload.lua script is a better place for automatic playlist management due to the Single Responsibility Principle

File: TOOLS/lua/autoload.lua [hunk 1]
```diff
@@ -67,9 +75,45 @@ function SetUnion (a,b)
...
 
-function Split (s)
+-- Returns first and last positions in string or past-to-end indices
+function FindOrPastTheEnd (string, pattern, start_at)
+    local pos1, pos2 = string.find(string, pattern, start_at)
+    return pos1 or #string + 1,
+           pos2 or #string + 1
+end
+
+function Split (list)
```
Comment (kasper93) [inline TOOLS/lua/autoload.lua:85]: not, function names are inconsistent. should be lower_case probably.

File: TOOLS/lua/autoload.lua [hunk 2]
```diff
@@ -177,9 +236,14 @@ function scan_dir(path, current_file, dir_mode, separator, dir_depth, total_file
     table.filter(files, function (v)
         -- The current file could be a hidden file, ignoring it doesn't load other
         -- files from the current directory.
-        if (o.ignore_hidden and not (prefix .. v == current_file) and string.match(v, "^%.")) then
+        local current = prefix .. v == current_file
+        if (o.ignore_hidden and not current and string.match(v, "^%.")) then
             return false
         end
+        if (not current and is_ignored(v)) then
```
Comment (kasper93) [inline TOOLS/lua/autoload.lua:243]: style, not parentheses

File: TOOLS/lua/autoload.lua (hunk 1 above)
Comment (oficsu) [reply TOOLS/lua/autoload.lua:85]: I followed the style of original Split (also Set/SetUnion above). Should I make them all lower_case?

Comment (kasper93) [reply TOOLS/lua/autoload.lua:85]: No, if there is precedence for this style, no need to bother. Can be changed separately.

File: TOOLS/lua/autoload.lua (hunk 2 above)
Comment (oficsu) [reply TOOLS/lua/autoload.lua:243]: Thanks, I'll fix this. Should I add a new commit to the PR? Or can I rebase & force-push PR?

Comment (kasper93) [reply TOOLS/lua/autoload.lua:243]: rebase and force push is ok

File: TOOLS/lua/autoload.lua [hunk 3]
```diff
@@ -22,6 +28,7 @@ additional_audio_exts=list,of,ext
 ignore_hidden=yes
 same_type=yes
 directory_mode=recursive
+ignore_patterns=^~,^bak-,%.bak%.mp4$
```
Comment (kasper93) [inline TOOLS/lua/autoload.lua:31]:
```suggestion
ignore_patterns=^~,^bak-,%.bak$
```
I would make the default more generic. `mp4` is quite specific and why not add all other media extensions? So I would just remove and leave `.bak`

**Pull request merged**